src/Database/LSMTree/Internal.hs

```diff
@@ -723,29 +705,261 @@ updatesWithInterleavedFlushes conf hfs root uniqC es reg tc = do
+        case mergePolicyForLevel confMergePolicy ln ls of
+          -- If r is still too small for this level then keep it and merge again
+          -- with the incoming runs.
+          LevelTiering | runSize r <= maxRunSize' conf LevelTiering (ln - 1) -> do
+            let mergelast = mergeLastForLevel ls
+            mr' <- newMerge LevelTiering mergelast (rs' `V.snoc` r)
+            pure $! Level mr' rs `V.cons` ls
+          -- This tiering level is now full. We take the completed merged run
+          -- (the previous incoming runs), plus all the other runs on this level
+          -- as a bundle and move them down to the level below. We start a merge
+          -- for the new incoming runs. This level is otherwise empty.
+          LevelTiering | levelIsFull rs -> do
+            mr' <- newMerge LevelTiering Merge.MidLevel rs'
+            ls' <- go (ln+1) (r `V.cons` rs) ls
+            pure $! Level mr' V.empty `V.cons` ls'
+          -- This tiering level is not yet full. We move the completed merged run
+          -- into the level proper, and start the new merge for the incoming runs.
+          LevelTiering -> do
+            let mergelast = mergeLastForLevel ls
+            mr' <- newMerge LevelTiering mergelast rs'
+            pure $! Level mr' (r `V.cons` rs) `V.cons` ls
+          -- The final level is using levelling. If the existing completed merge
+          -- run is too large for this level, we promote the run to the next
+          -- level and start merging the incoming runs into this (otherwise
+          -- empty) level .
+          LevelLevelling | runSize r > maxRunSize' conf LevelLevelling ln -> do
+            assert (V.null rs && V.null ls) $ pure ()
+            mr' <- newMerge LevelTiering Merge.MidLevel rs'
+            ls' <- go (ln+1) (V.singleton r) V.empty
+            pure $! Level mr' V.empty `V.cons` ls'
+          -- Otherwise we start merging the incoming runs into the run.
+          LevelLevelling -> do
+            assert (V.null rs && V.null ls) $ pure ()
+            mr' <- newMerge LevelLevelling Merge.LastLevel (rs' `V.snoc` r)
+            pure $! Level mr' V.empty `V.cons` V.empty
+
+    expectCompletedMerge :: MergingRun (Handle h) -> m (Run (Handle h))
+    expectCompletedMerge (SingleRun r)                   = pure r
+    expectCompletedMerge (MergingRun (CompletedMerge r)) = pure r
+
+    -- TODO: Until we implement proper scheduling, this does not only start a
+    -- merge, but it also steps it to completion.
+    newMerge :: MergePolicyForLevel -> Merge.Level -> V.Vector (Run (Handle h)) -> m (MergingRun (Handle h))
+    newMerge mergepolicy mergelast rs
+      | Just (r, rest) <- V.uncons rs
+      , V.null rest = do
+          pure (SingleRun r)
+      | otherwise = do
+        assert (let l = V.length rs in l >= 2 && l <= 5) $ pure ()
+        r <- allocateTemp reg (mergeRuns conf hfs root uniqC mergepolicy mergelast rs) (Run.removeReference hfs)
```

Comment:
This is very coarse grained masking: the whole merge is masked! If we want more coarse grained masking, then we'll have to thread the registry through merges and flushes. However, it might not be worth it since we are going to schedule merges and then this code has to change anyway